File: tools/core070_bench_full.jl
# M3 full benchmark campaign, Julia side. Reads bench-manifest.csv +
# bench-Y-*.csv from the R stage's output dir; fits the paired canonical
# model natively; 5 reps, median wall seconds.
# Family drivers: fit_gaussian_gllvm (trait-dummy X tensor, ns2 convention),
# fit_poisson_gllvm / fit_nb_gllvm / fit_binomial_gllvm (intrinsic per-trait
# beta intercepts = R's 0 + trait means).
# argv: ARGS[1] = the campaign output dir
using GLLVM
using DelimitedFiles
using Statistics

out = ARGS[1]
man, _ = readdlm(joinpath(out, "bench-manifest.csv"), ',', header = true)
_loglik(fit) = hasproperty(fit, :logLik) ? fit.logLik : fit.loglik

open(joinpath(out, "julia-timings.csv"), "w") do io
    println(io, "case,p,n,family,median_s,min_s,loglik,ok")
    for i in axes(man, 1)
        case = Int(man[i, 1]); p = Int(man[i, 2]); n = Int(man[i, 3])
        fam = String(strip(string(man[i, 4]), ['"', ' ']))
        raw, _ = readdlm(joinpath(out, "bench-Y-" * lpad(case, 3, '0') * ".csv"),
                         ',', header = true)
        Y = Matrix{Float64}(raw)
        @assert size(Y) == (p, n)
        times = Float64[]; ll = NaN; ok = true
        for r in 1:5
            t = @elapsed try
                fit = if fam == "gaussian"
                    X = zeros(p, n, p)
                    for j in 1:p
                        X[j, :, j] .= 1
                    end
                    GLLVM.fit_gaussian_gllvm(Y; K = 2, X = X)
                elseif fam == "poisson"
                    GLLVM.fit_poisson_gllvm(Int.(round.(Y)); K = 2)
                elseif fam == "nbinom2"
                    GLLVM.fit_nb_gllvm(Int.(round.(Y)); K = 2)
                elseif fam == "binomial"
                    GLLVM.fit_binomial_gllvm(Int.(round.(Y)); K = 2)
                else
                    error("unknown family $fam")
                end
                ll = _loglik(fit)
            catch err
                ok = false
                println(stderr, "case=$case $fam FAILED: ", sprint(showerror, err)[1:200])
            end
            push!(times, t)
            ok || break
        end
        println(io, join([case, p, n, fam, median(times), minimum(times), ll, ok], ","))
        flush(io)
        println(stderr, "J done case=$case $fam p=$p n=$n median=$(round(median(times), digits = 3))s ok=$ok")
    end
end
println("BENCH_FULL_JULIA_DONE")

File: tools/core070_bench_full.R
# M3 full benchmark campaign, R side (maintainer-approved 2026-09-01).
# Families: gaussian, poisson, nbinom2, binomial(logit; Bernoulli cells).
# Grid: p in {5,10,20,50} x n in {200,500,1000} plus (100,200) and (100,500);
# 5 reps, median wall time, single thread. Paired canonical model
# (0 + trait means, d=2 latent, unique=FALSE; n_init=1, se off).
# argv: <frozen-library> <output-dir>
args <- commandArgs(TRUE)
stopifnot(length(args) == 2L)
frozen_library <- normalizePath(args[[1]], mustWork = TRUE)
out_dir <- args[[2]]
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
.libPaths(c(frozen_library, .libPaths()))
suppressMessages(library(gllvmTMB))

cells <- rbind(expand.grid(p = c(5L, 10L, 20L, 50L), n = c(200L, 500L, 1000L)),
               data.frame(p = 100L, n = c(200L, 500L)))
fams <- c("gaussian", "poisson", "nbinom2", "binomial")
grid <- merge(cells, data.frame(family = fams))
reps <- 5L
rows <- list()
for (g in seq_len(nrow(grid))) {
  p <- grid$p[g]; n <- grid$n[g]; fam <- grid$family[g]
  set.seed(91000 + g)
  K <- 2L
  Lambda <- matrix(rnorm(p * K, 0, 0.5), p, K)
  eta <- Lambda %*% matrix(rnorm(K * n), K, n)
  Y <- switch(fam,
    gaussian = eta + 0.7 * matrix(rnorm(p * n), p, n),
    poisson  = matrix(rpois(p * n, exp(pmin(eta, 4))), p, n),
    nbinom2  = matrix(rnbinom(p * n, mu = exp(pmin(eta, 4)), size = 2), p, n),
    binomial = matrix(rbinom(p * n, 1L, plogis(eta)), p, n))
  tn <- paste0("t", seq_len(p))
  df <- data.frame(site = factor(rep(seq_len(n), each = p)),
                   trait = factor(rep(tn, n), levels = tn),
                   value = as.vector(Y))
  fam_obj <- switch(fam, gaussian = gaussian(), poisson = poisson(),
                    nbinom2 = nbinom2(), binomial = binomial())
  times <- rep(NA_real_, reps); ll <- NA_real_; ok <- TRUE
  for (r in seq_len(reps)) {
    t0 <- proc.time()[["elapsed"]]
    fit <- try(gllvmTMB(value ~ 0 + trait + latent(0 + trait | site, d = 2, unique = FALSE),
                        data = df, unit = "site", trait = "trait", family = fam_obj,
                        control = gllvmTMBcontrol(n_init = 1L, se = FALSE)), silent = TRUE)
    times[r] <- proc.time()[["elapsed"]] - t0
    if (inherits(fit, "try-error")) { ok <- FALSE; break }
    ll <- as.numeric(logLik(fit))
  }
  rows[[length(rows) + 1L]] <- data.frame(case = g, p = p, n = n, family = fam,
    median_s = median(times, na.rm = TRUE), min_s = min(times, na.rm = TRUE),
    loglik = ll, ok = ok)
  write.csv(Y, file.path(out_dir, sprintf("bench-Y-%03d.csv", g)), row.names = FALSE)
  cat(sprintf("R done %d/%d %s p=%d n=%d median=%.2fs ok=%s\n",
              g, nrow(grid), fam, p, n, median(times, na.rm = TRUE), ok))
}
res <- do.call(rbind, rows)
res$engine <- "gllvmTMB-frozen"
write.csv(res, file.path(out_dir, "r-timings.csv"), row.names = FALSE)
write.csv(data.frame(case = seq_len(nrow(grid)), p = grid$p, n = grid$n,
                     family = grid$family),
          file.path(out_dir, "bench-manifest.csv"), row.names = FALSE)
